src/api/procedures/issueTokens.ts

```diff
@@ -27,49 +26,47 @@ export async function prepareSetIssuancesData(
     },
     context,
   } = this;
-  const { ticker, issuances } = args;
+  const { ticker, issuanceData } = args;
 
   const securityToken = new SecurityToken({ ticker }, context);
 
   const { isDivisible, totalSupply } = await securityToken.details();
-  const values = issuances.map(issuance => new BigNumber(issuance.balance));
+  const values = issuanceData.map(issuance => issuance.amount);
 
   values.forEach(value => {
     if (isDivisible) {
-      if (value.decimalPlaces() > 6) {
+      if (value.decimalPlaces() > MAX_DECIMALS) {
         throw new PolymeshError({
           code: ErrorCode.ValidationError,
-          message: 'At most one balance exceeds the six decimals limit',
+          message: `Issuance amounts cannot have more than ${MAX_DECIMALS} decimals`,
         });
       }
     } else {
       if (value.decimalPlaces()) {
         throw new PolymeshError({
           code: ErrorCode.ValidationError,
-          message: 'At most one balance has decimals',
+          message: 'Cannot issue decimal amounts of an indivisible token',
         });
       }
     }
   });
 
-  const mintSupply = values.reduce((acc, next) => {
+  const supplyAfterMint = values.reduce((acc, next) => {
     return acc.plus(next);
-  }, new BigNumber(0));
-
-  const limitTotalSupply = new BigNumber(Math.pow(10, 12));
+  }, totalSupply);
 
-  if (totalSupply.plus(mintSupply).isGreaterThan(limitTotalSupply)) {
+  if (supplyAfterMint.isGreaterThan(MAX_TOKEN_AMOUNT)) {
     throw new PolymeshError({
       code: ErrorCode.ValidationError,
-      message: `The total supply for "${ticker}" cannot be bigger than ${limitTotalSupply.toString()}`,
+      message: `This issuance operation will cause the total supply of "${ticker}" to exceed the maximum allowed (${MAX_TOKEN_AMOUNT.toFormat()})`,
     });
   }
 
   // TODO: check canTransfer method
 
   const rawTicker = stringToTicker(ticker, context);
-  const investors = issuances.map(issuance => stringToIdentityId(issuance.did, context));
-  const balances = issuances.map(issuance => numberToBalance(issuance.balance, context));
+  const investors = issuanceData.map(issuance => stringToIdentityId(issuance.did, context));
+  const balances = issuanceData.map(issuance => numberToBalance(issuance.amount, context));
```

Comment:
You should create arrays and fill them in a forEach, it's a tiny bit more efficient